[LTO] [MinGW] Use of thread locals from different translation unit broken with LTO

Hi all,

This is to report that if both `-flto` and `-fopenmp` are enabled for a target, and it makes use of a static library that contains a symbol exported by `#pragma openmp threadprivate(...)`, linking will fail with the following error:

```
FAILED: subprojects/pixman-0.42.2/test/region-test.exe
"cc"  -o subprojects/pixman-0.42.2/test/region-test.exe subprojects/pixman-0.42.2/test/region-test.exe.p/region-test.c.obj "-flto" "-Wl,--start-group" "subprojects\pixman-0.42.2\test\libtestutils.a" "subprojects\pixman-0.42.2\pixman\libpixman-1.a" "subprojects\pixman-0.42.2\pixman\libpixman-mmx.a" "subprojects\pixman-0.42.2\pixman\libpixman-sse2.a" "subprojects\pixman-0.42.2\pixman\libpixman-ssse3.a" "-pthread" "-fopenmp" "-fopenmp" "-lm" "D:/msys64/clang64/bin/../lib/libpng16.dll.a" "-Wl,--subsystem,console" "-lkernel32" "-luser32" "-lgdi32" "-lwinspool" "-lshell32" "-lole32" "-loleaut32" "-luuid" "-lcomdlg32" "-ladvapi32" "-Wl,--end-group"
ld.lld: error: undefined symbol: thread-local initialization routine for prng_state
>>> referenced by ../subprojects/pixman-0.42.2/test/region-test.c
>>>               subprojects/pixman-0.42.2/test/region-test.exe.p/region-test.c.obj

ld.lld: error: undefined symbol: thread-local initialization routine for prng_state_data
>>> referenced by ../subprojects/pixman-0.42.2/test/region-test.c
>>>               subprojects/pixman-0.42.2/test/region-test.exe.p/region-test.c.obj
cc: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Linking with `-flto=thin`, as well as building with GCC + `-flto`, both result in a working build.

To reproduce, please download https://www.cairographics.org/releases/pixman-0.42.2.tar.gz, unpack it, and then perform:

```
cd pixman-0.42.2
meson setup build -Ddefault_library=static -Db_lto=true -Dbackend_max_links=1
meson compile -C build
```

clang version 16.0.5
Target: x86_64-w64-windows-gnu
Thread model: posix
InstalledDir: D:/msys64/clang64/bin

--- a/lld/COFF/InputFiles.cpp
+++ b/lld/COFF/InputFiles.cpp
@@ -1040,6 +1040,8 @@ void BitcodeFile::parse() {
       fakeSC = &ctx.ltoDataSectionChunk.chunk;
     if (objSym.isUndefined()) {
       sym = ctx.symtab.addUndefined(symName, this, false);
+      if (objSym.isWeak())
+        sym->deferUndefined = true;
     } else if (objSym.isCommon()) {
       sym = ctx.symtab.addCommon(this, symName, objSym.getCommonSize());
     } else if (objSym.isWeak() && objSym.isIndirect()) {

--- a/lld/test/COFF/lto-weak-undefined.ll
+++ b/lld/test/COFF/lto-weak-undefined.ll
@@ -1,0 +1,40 @@
+; REQUIRES: x86
+
+;; Test linking of weak symbols with LTO. The weak symbol may be defined
+;; by another object file, or may be left undefined. When compiling the
+;; IR with an undefined weak symbol, the emitted object file will contain
+;; a weak alias pointing at an absolute symbol for the address null.
+;; Make sure both cases can be linked correctly.
+
+; RUN: split-file %s %t.dir
+; RUN: llvm-as %t.dir/main.ll -o %t.main.obj
+; RUN: llvm-as %t.dir/optional.ll -o %t.optional.obj
+
+; RUN: lld-link /entry:main %t.main.obj /out:%t-undef.exe
+; RUN: lld-link /entry:main %t.main.obj %t.optional.obj /out:%t-def.exe
+
+;--- main.ll
+target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-windows-msvc"
+
+define dso_local i32 @main() {
+entry:
+  br i1 icmp ne (ptr @optionalFunc, ptr null), label %if.then, label %if.end
+
+if.then:
+  tail call void @optionalFunc()
+  br label %if.end
+
+if.end:
+  ret i32 0
+}
+
+declare extern_weak void @optionalFunc()
+
+;--- optional.ll
+target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-windows-msvc"
+
+define void @optionalFunc() {
+  ret void
+}

